lib/pos_blockchain.go

```diff
@@ -374,17 +366,144 @@ func (bc *Blockchain) getLineageFromCommittedTip(desoBlock *MsgDeSoBlock) ([]*Bl
+	// If the BlockNode is not already stored, we should set its status to stored.
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	return bc.upsertBlockToDBBlockIndex(desoBlock, blockNode)
+}
+
+// storeValidateFailedBlockInBlockIndex upserts the blocks into the in-memory block index and updates its status to
+// StatusBlockValidateFailed. If it does not have the status StatusBlockStored already, we add that as we will
+// store the block in the DB after updating its status.  It also writes the block to the block index in badger
+// by calling upsertBlockToDBBlockIndex.
+func (bc *Blockchain) storeValidateFailedBlockInBlockIndex(desoBlock *MsgDeSoBlock) error {
+	blockNode, err := bc.getOrCreateBlockNodeFromBlockIndex(desoBlock)
+	if err != nil {
+		return errors.Wrapf(err, "storeValidateFailedBlockInBlockIndex: Problem getting or creating block node")
+	}
+	// We should throw an error if the BlockNode is already Validate Failed.
+	if blockNode.IsValidateFailed() {
+		return errors.New("storeValidateFailedBlockInBlockIndex: BlockNode is already validate failed")
+	}
+	// We should throw an error if the BlockNode is already Validated
+	if blockNode.IsValidated() {
+		return errors.New("storeValidateFailedBlockInBlockIndex: can't set BlockNode to validate failed after it's already validated")
+	}
+	blockNode.Status |= StatusBlockValidateFailed
+	// If the BlockNode is not already stored, we should set it to stored.
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	return bc.upsertBlockToDBBlockIndex(desoBlock, blockNode)
+}
+
+// storeCommittedBlockInBlockIndex upserts the blocks into the in-memory block index and updates its status to
+// StatusBlockCommitted. If the BlockNode does not have StatusBlockValidated and StatusBlockStored statuses,
+// we also add those. It also writes the block to the block index in badger by calling upsertBlockToDBBlockIndex.
+func (bc *Blockchain) storeCommittedBlockInBlockIndex(desoBlock *MsgDeSoBlock) error {
+	blockNode, err := bc.getOrCreateBlockNodeFromBlockIndex(desoBlock)
+	if err != nil {
+		return errors.Wrapf(err, "storeCommittedBlockInBlockIndex: Problem getting or creating block node")
+	}
+	// We should throw an error if the BlockNode is already Committed.
+	if blockNode.IsCommitted() {
+		return errors.New("storeCommittedBlockInBlockIndex: BlockNode is already committed")
+	}
+	blockNode.Status |= StatusBlockCommitted
+	if !blockNode.IsValidated() {
+		blockNode.Status |= StatusBlockValidated
+	}
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	// Do we want any side effects here or nah?
+	return bc.upsertBlockToDBBlockIndex(desoBlock, blockNode)
+}
+
+// upsertBlockToDBBlockIndex adds the block to the block index with the given status. It also writes the block to the block
+// index in badger.
+func (bc *Blockchain) upsertBlockToDBBlockIndex(desoBlock *MsgDeSoBlock, newBlockNode *BlockNode) error {
```

Comment:
WDYT of this name? Seems a bit clearer since the in-memory block index isn't being updated, but rather the block and its node are being persisted to DB
```suggestion
func (bc *Blockchain) upsertBlockAndBlockNodeToDB(block *MsgDeSoBlock, blockNode *BlockNode) error {
```